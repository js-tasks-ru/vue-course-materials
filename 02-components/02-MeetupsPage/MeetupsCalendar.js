const MeetupsCalendar = {
  name: 'MeetupsCalendar',

  props: {
    meetups: {
      type: Array,
      required: true,
    },
  },

  template: `<div>Календарь</div>`,
};

export default MeetupsCalendar;
