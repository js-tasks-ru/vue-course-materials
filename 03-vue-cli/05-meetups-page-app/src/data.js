/** URL адрес API */
export const API_URL = process.env.VUE_APP_API_URL;

/**
 * Возвращает ссылку на изображение по идентификатору, например, изображение митапа
 * @param imageId {number} - идентификатор изображения
 * @return {string} - ссылка на изображение
 */
export function getImageUrlByImageId(imageId) {
  return `${API_URL}/images/${imageId}`;
}

/**
 * Получает список митапов с API
 * @return {Promise}
 */
export async function fetchMeetups() {
  return fetch(`${API_URL}/meetups`).then((res) => res.json());
}
