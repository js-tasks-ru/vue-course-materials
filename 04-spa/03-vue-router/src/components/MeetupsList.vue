<template>
  <div class="meetups-list">
    <meetup-list-item v-for="meetup in meetups" :meetup="meetup" :key="meetup.id" />
  </div>
</template>

<script>
import MeetupListItem from './MeetupsListItem';

export default {
  name: 'MeetupsList',

  props: {
    meetups: {
      type: Array,
      required: true,
    },
  },

  components: {
    MeetupListItem,
  },
};
</script>

<style scoped>
.meetups-list {
}
</style>
