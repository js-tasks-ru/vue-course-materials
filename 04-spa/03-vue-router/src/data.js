/** URL адрес API */
export const API_URL = process.env.VUE_APP_API_URL;

/**
 * Возвращает ссылку на изображение митапа для митапа
 * @param meetup - объект с описанием митапа (и параметром meetupId)
 * @return {string} - ссылка на изображение митапа
 */
export function getMeetupCoverLink(meetup) {
  return `${API_URL}/images/${meetup.imageId}`;
}

/**
 * Получает список митапов с API
 * @return {Promise}
 */
export async function fetchMeetups() {
  return fetch(`${API_URL}/meetups`).then((res) => res.json());
}
