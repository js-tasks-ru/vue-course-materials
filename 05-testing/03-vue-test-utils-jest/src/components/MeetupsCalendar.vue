<template>
  <div class="meetups-calendar">Calendar</div>
</template>

<script>
export default {
  name: 'MeetupsCalendar',

  props: {
    meetups: {
      type: Array,
      required: true,
    },
  },
};
</script>

<style scoped></style>
