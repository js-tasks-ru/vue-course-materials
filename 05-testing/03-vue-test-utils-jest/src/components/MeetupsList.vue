<template>
  <div class="meetups-list">
    <meetup-list-item
      v-for="meetup in meetups"
      :meetup="meetup"
      :key="meetup.id"
    ></meetup-list-item>
  </div>
</template>

<script>
import MeetupListItem from './MeetupsListItem';

export default {
  name: 'MeetupsList',

  props: {
    meetups: {
      type: Array,
      required: true,
    },
  },

  components: {
    MeetupListItem,
  },
};
</script>

<style scoped></style>
