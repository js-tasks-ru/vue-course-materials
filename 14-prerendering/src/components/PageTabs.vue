<template>
  <div class="page-tabs">
    <button
      type="button"
      class="page-tabs__tab"
      :class="{
        'page-tabs__tab_active': selected === '' || selected === 'list',
      }"
      @click="select('list')"
    >
      <svg
        fill="none"
        height="28"
        viewBox="0 0 28 28"
        width="28"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="m5 7c-.55228 0-1 .44772-1 1s.44772 1 1 1h.01c.55228 0 1-.44772 1-1s-.44772-1-1-1zm5 0c-.55228 0-1 .44772-1 1s.44772 1 1 1h13c.5523 0 1-.44772 1-1s-.4477-1-1-1zm0 6c-.55228 0-1 .4477-1 1s.44772 1 1 1h13c.5523 0 1-.4477 1-1s-.4477-1-1-1zm-1 7c0-.5523.44772-1 1-1h13c.5523 0 1 .4477 1 1s-.4477 1-1 1h-13c-.55228 0-1-.4477-1-1zm-5-6c0-.5523.44772-1 1-1h.01c.55228 0 1 .4477 1 1s-.44772 1-1 1h-.01c-.55228 0-1-.4477-1-1zm1 5c-.55228 0-1 .4477-1 1s.44772 1 1 1h.01c.55228 0 1-.4477 1-1s-.44772-1-1-1z"
          fill-rule="evenodd"
        ></path>
      </svg>
    </button>
    <button
      type="button"
      class="page-tabs__tab"
      :class="{ 'page-tabs__tab_active': selected === 'calendar' }"
      @click="select('calendar')"
    >
      <svg
        height="22"
        viewBox="0 0 20 22"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="m15 1c0-.552285-.4477-1-1-1s-1 .447715-1 1v1h-6v-1c0-.552285-.44772-1-1-1s-1 .447715-1 1v1h-2c-1.65685 0-3 1.34315-3 3v14c0 1.6569 1.34315 3 3 3h14c1.6569 0 3-1.3431 3-3v-14c0-1.65685-1.3431-3-3-3h-2zm3 7v-3c0-.55228-.4477-1-1-1h-2v1c0 .55228-.4477 1-1 1s-1-.44772-1-1v-1h-6v1c0 .55228-.44772 1-1 1s-1-.44772-1-1v-1h-2c-.55228 0-1 .44772-1 1v3zm-16 2h16v9c0 .5523-.4477 1-1 1h-14c-.55228 0-1-.4477-1-1z"
          fill-rule="evenodd"
        ></path>
      </svg>
    </button>
  </div>
</template>

<script>
export default {
  name: 'PageTabs',

  props: {
    selected: String,
  },

  methods: {
    select(name) {
      this.$emit('update:selected', name);
    },
  },
};
</script>

<style scoped></style>
