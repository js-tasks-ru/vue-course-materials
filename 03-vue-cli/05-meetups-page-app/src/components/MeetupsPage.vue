<template>
  <div class="container">
    <div class="filters-panel">
      <div class="filters-panel__col">
        <form-check v-model="filter.date" :options="$options.dateFilterOptions" name="date" />
      </div>

      <div class="filters-panel__col">
        <div class="form-group form-group_inline">
          <div class="input-group input-group_icon input-group_icon-left">
            <app-icon icon="search" />
            <input
              id="filters-panel__search"
              class="form-control form-control_rounded form-control_sm"
              type="text"
              placeholder="Поиск"
              v-model="filter.search"
            />
          </div>
        </div>
        <div class="form-group form-group_inline">
          <page-tabs :selected.sync="view" />
        </div>
      </div>
    </div>

    <template v-if="filteredMeetups && filteredMeetups.length">
      <meetups-list v-if="view === 'list'" :meetups="filteredMeetups" key="list"></meetups-list>
      <meetups-calendar v-else-if="view === 'calendar'" :meetups="filteredMeetups" key="calendar"></meetups-calendar>
    </template>
    <app-empty v-else>Митапов по заданным условиям не найдено...</app-empty>
  </div>
</template>

<script>
import { fetchMeetups, getImageUrlByImageId } from '@/data';
import MeetupsList from '@/components/MeetupsList';
import MeetupsCalendar from '@/components/MeetupsCalendar';
import PageTabs from '@/components/PageTabs';
import FormCheck from '@/components/FormCheck';
import AppEmpty from '@/components/AppEmpty';
import AppIcon from '@/components/AppIcon';

export default {
  name: 'MeetupsPage',

  dateFilterOptions: [
    { text: 'Все', value: 'all' },
    { text: 'Прошедшие', value: 'past' },
    { text: 'Ожидаемые', value: 'future' },
  ],

  components: {
    MeetupsList,
    MeetupsCalendar,
    PageTabs,
    FormCheck,
    AppEmpty,
    AppIcon,
  },

  data() {
    return {
      rawMeetups: [],
      filter: {
        date: 'all',
        participation: 'all',
        search: '',
      },
      view: 'list',
    };
  },

  mounted() {
    this.fetchMeetups();
  },

  computed: {
    meetups() {
      return this.rawMeetups.map((meetup) => ({
        ...meetup,
        cover: meetup.imageId ? getImageUrlByImageId(meetup.imageId) : undefined,
        coverStyle: meetup.imageId
          ? {
              '--bg-url': `url('${getImageUrlByImageId(meetup.imageId)}')`,
            }
          : {},
        date: new Date(meetup.date),
        localDate: new Date(meetup.date).toLocaleString(navigator.language, {
          year: 'numeric',
          month: 'long',
          day: 'numeric',
        }),
        dateOnlyString: new Date(meetup.date).toISOString().split('T'),
      }));
    },

    filteredMeetups() {
      const dateFilter = (meetup) =>
        this.filter.date === 'all' ||
        (this.filter.date === 'past' && new Date(meetup.date) <= new Date()) ||
        (this.filter.date === 'future' && new Date(meetup.date) > new Date());

      const participationFilter = (meetup) =>
        this.filter.participation === 'all' ||
        (this.filter.participation === 'organizing' && meetup.organizing) ||
        (this.filter.participation === 'attending' && meetup.attending);

      const searchFilter = (meetup) =>
        [meetup.title, meetup.description, meetup.place, meetup.organizer]
          .join(' ')
          .toLowerCase()
          .includes(this.filter.search.toLowerCase());

      return this.meetups.filter((meetup) => dateFilter(meetup) && participationFilter(meetup) && searchFilter(meetup));
    },
  },

  methods: {
    async fetchMeetups() {
      this.rawMeetups = await fetchMeetups();
    },
  },
};
</script>

<style scoped>
.filters-panel {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  margin: 40px 0 32px 0;
}

.filters-panel__col {
  display: flex;
  flex-direction: row;
  align-items: flex-start;
}

@media all and (max-width: 767px) {
  .filters-panel {
    flex-direction: column;
  }

  .filters-panel__col + .filters-panel__col {
    margin-top: 16px;
  }
}

@media all and (max-width: 480px) {
  #filters-panel__search {
    width: calc(100% - 112px);
  }

  #filters-panel__search >>> .form-control {
    max-width: 100%;
  }
}
</style>
