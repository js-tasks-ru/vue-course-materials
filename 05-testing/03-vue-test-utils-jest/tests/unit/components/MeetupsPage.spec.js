import flushPromises from 'flush-promises';
import { shallowMount, mount } from '@vue/test-utils';
import MeetupsPage from '@/components/MeetupsPage';
import MeetupsList from '@/components/MeetupsList';
import { fetchMeetups } from '@/data';
import meetups from '../__fixtures__/meetups.json';
import PageTabs from '@/components/PageTabs';
import MeetupsCalendar from '@/components/MeetupsCalendar';

jest.mock('@/data');

describe('MeetupsPage', () => {
  beforeAll(() => {
    fetchMeetups.mockResolvedValue(meetups);
  });

  it('should be defined', () => {
    expect(MeetupsPage).toBeDefined();
  });

  describe('Unit testing', () => {
    it('should render list with 2 meetups received from fetchMeetups by <meetups-list>', async () => {
      const wrapper = mount(MeetupsPage, {
        stubs: {
          AppIcon: true,
          AppEmpty: true,
          MeetupsCalendar: true,
          PageTabs: true,
          FormCheck: true,
          AppInput: true,
          MeetupsList: {
            props: ['meetups'],
            render(h) {
              return h(
                'div',
                this.meetups.map((meetup) => meetup.title).join(' '),
              );
            },
          },
        },
      });
      await flushPromises();
      expect(fetchMeetups).toHaveBeenCalled();
      expect(wrapper.findComponent(MeetupsList).props('meetups')).toMatchObject(
        meetups,
      );
    });

    it('should render list with 2 meetups received from fetchMeetups by <meetups-list> (shallowMount)', async () => {
      const wrapper = shallowMount(MeetupsPage);
      await flushPromises();
      expect(fetchMeetups).toHaveBeenCalled();
      expect(wrapper.findComponent(MeetupsList).props('meetups')).toMatchObject(
        meetups,
      );
    });

    it('should render MeetupsCalendar instead of MeetupsList after select calendar by PageTabs', async () => {
      const wrapper = shallowMount(MeetupsPage);
      await flushPromises();
      const pageTabs = wrapper.findComponent(PageTabs);
      pageTabs.vm.$emit('update:selected', 'calendar');
      await wrapper.vm.$nextTick();
      expect(wrapper.findComponent(MeetupsList).exists()).toBeFalsy();
      expect(wrapper.findComponent(MeetupsCalendar).exists()).toBeTruthy();
    });

    it('should match shallowMount snapshot', async () => {
      const wrapper = shallowMount(MeetupsPage);
      await flushPromises();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Integration testing', () => {
    it('should render list with 2 meetups titles received from fetchMeetups in .meetups-list__cover', async () => {
      const wrapper = mount(MeetupsPage);
      await flushPromises();
      expect(fetchMeetups).toHaveBeenCalled();
      expect(
        wrapper
          .findAll('.meetups-list__cover')
          .wrappers.map((wrapper) => wrapper.text()),
      ).toEqual(meetups.map((meetup) => meetup.title));
    });

    it('should render .meetups-calendar instead of .meetups-list after click on second .page-tabs__tab button', async () => {
      const wrapper = mount(MeetupsPage);
      await flushPromises();
      const pageTabs = wrapper.findAll('.page-tabs__tab').wrappers;
      await pageTabs[1].trigger('click');
      expect(wrapper.find('.meetups-list').exists()).toBeFalsy();
      expect(wrapper.find('.meetups-calendar').exists()).toBeTruthy();
    });

    it('should match mount snapshot', async () => {
      const wrapper = mount(MeetupsPage);
      await flushPromises();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
