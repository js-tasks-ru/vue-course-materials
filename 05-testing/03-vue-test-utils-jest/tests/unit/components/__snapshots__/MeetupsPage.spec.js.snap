// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeetupsPage Integration testing should match mount snapshot 1`] = `
<div class="container">
  <div class="filters-panel">
    <div class="filters-panel__col">
      <div class="form-check">
        <div class="form-check__group"><input type="radio" name="date" class="form-check__input" value="all"> <label class="form-check__label">Все</label></div>
        <div class="form-check__group"><input type="radio" name="date" class="form-check__input" value="past"> <label class="form-check__label">Прошедшие</label></div>
        <div class="form-check__group"><input type="radio" name="date" class="form-check__input" value="future"> <label class="form-check__label">Ожидаемые</label></div>
      </div>
    </div>
    <div class="filters-panel__col">
      <div class="form-group form-group_inline">
        <div class="input-group input-group_icon input-group_icon-left"><img alt="icon" src="" class="icon"> <input id="filters-panel__search" type="text" placeholder="Поиск" class="form-control form-control_rounded form-control_sm"></div>
      </div>
      <div class="form-group form-group_inline">
        <div class="page-tabs"><button type="button" class="page-tabs__tab page-tabs__tab_active"><svg fill="none" height="28" viewBox="0 0 28 28" width="28" xmlns="http://www.w3.org/2000/svg">
              <path clip-rule="evenodd" d="m5 7c-.55228 0-1 .44772-1 1s.44772 1 1 1h.01c.55228 0 1-.44772 1-1s-.44772-1-1-1zm5 0c-.55228 0-1 .44772-1 1s.44772 1 1 1h13c.5523 0 1-.44772 1-1s-.4477-1-1-1zm0 6c-.55228 0-1 .4477-1 1s.44772 1 1 1h13c.5523 0 1-.4477 1-1s-.4477-1-1-1zm-1 7c0-.5523.44772-1 1-1h13c.5523 0 1 .4477 1 1s-.4477 1-1 1h-13c-.55228 0-1-.4477-1-1zm-5-6c0-.5523.44772-1 1-1h.01c.55228 0 1 .4477 1 1s-.44772 1-1 1h-.01c-.55228 0-1-.4477-1-1zm1 5c-.55228 0-1 .4477-1 1s.44772 1 1 1h.01c.55228 0 1-.4477 1-1s-.44772-1-1-1z" fill-rule="evenodd"></path>
            </svg></button> <button type="button" class="page-tabs__tab"><svg height="22" viewBox="0 0 20 22" width="20" xmlns="http://www.w3.org/2000/svg">
              <path clip-rule="evenodd" d="m15 1c0-.552285-.4477-1-1-1s-1 .447715-1 1v1h-6v-1c0-.552285-.44772-1-1-1s-1 .447715-1 1v1h-2c-1.65685 0-3 1.34315-3 3v14c0 1.6569 1.34315 3 3 3h14c1.6569 0 3-1.3431 3-3v-14c0-1.65685-1.3431-3-3-3h-2zm3 7v-3c0-.55228-.4477-1-1-1h-2v1c0 .55228-.4477 1-1 1s-1-.44772-1-1v-1h-6v1c0 .55228-.44772 1-1 1s-1-.44772-1-1v-1h-2c-.55228 0-1 .44772-1 1v3zm-16 2h16v9c0 .5523-.4477 1-1 1h-14c-.55228 0-1-.4477-1-1z" fill-rule="evenodd"></path>
            </svg></button></div>
      </div>
    </div>
  </div>
  <div class="meetups-list"><a href="/meetups/1" class="meetups-list__item">
      <div class="meetups-list__col">
        <div class="meetups-list__cover" style="--bg-url: url('https://course-vue.javascript.ru/api/images/1');">
          <h5>MSK VUE.JS MEETUP #1</h5>
        </div>
      </div>
      <div class="meetups-list__col">
        <div class="meetups-list__description">
          <!---->
          <!---->
          <ul class="info-list">
            <li><img alt="icon" src="" class="icon info-list__icon">
              Игорь Ш.
            </li>
            <li><img alt="icon" src="" class="icon info-list__icon">
              Москва, офис Voximplant (ул. Мытная 66)
            </li>
            <li><img alt="icon" src="" class="icon info-list__icon"> <time datetime="2020-05-08,00:00:00.000Z">May 8, 2020</time></li>
          </ul>
        </div>
      </div>
    </a><a href="/meetups/2" class="meetups-list__item">
      <div class="meetups-list__col">
        <div class="meetups-list__cover">
          <h5>Демо-Митап</h5>
        </div>
      </div>
      <div class="meetups-list__col">
        <div class="meetups-list__description">
          <!---->
          <!---->
          <ul class="info-list">
            <li><img alt="icon" src="" class="icon info-list__icon">
              Demo Organizer
            </li>
            <li><img alt="icon" src="" class="icon info-list__icon">
              Internet
            </li>
            <li><img alt="icon" src="" class="icon info-list__icon"> <time datetime="2020-07-20,16:15:59.010Z">July 20, 2020</time></li>
          </ul>
        </div>
      </div>
    </a></div>
</div>
`;

exports[`MeetupsPage Unit testing should match shallowMount snapshot 1`] = `
<div class="container">
  <div class="filters-panel">
    <div class="filters-panel__col">
      <form-check-stub options="[object Object],[object Object],[object Object]" value="all"></form-check-stub>
    </div>
    <div class="filters-panel__col">
      <div class="form-group form-group_inline">
        <div class="input-group input-group_icon input-group_icon-left">
          <app-icon-stub icon="search"></app-icon-stub> <input id="filters-panel__search" type="text" placeholder="Поиск" class="form-control form-control_rounded form-control_sm">
        </div>
      </div>
      <div class="form-group form-group_inline">
        <page-tabs-stub selected="list"></page-tabs-stub>
      </div>
    </div>
  </div>
  <meetups-list-stub meetups="[object Object],[object Object]"></meetups-list-stub>
</div>
`;
